File: core/sentinel.py
"""
Hallucination / anomaly sentinel + hot-swap failover.

One sentinel watches every worker agent. Each pipeline outcome for a lead
that agent handled is scored as anomalous or not (schema violation, a
reviewer catching a bid under the COGS floor, a malformed extraction --
anything core.reviewer/core.pricing already flags as invalid). The
sentinel keeps an EWMA drift score per agent; three consecutive anomalies
triggers a simulated hot-swap: the agent is quarantined, its state is
handed to a freshly spawned replacement, and the event is logged so an
admin panel (or, here, the API/UI) can show exactly what happened and why.
"""
from __future__ import annotations

from core.models import AgentNode, AgentStatus, SentinelEvent, new_id

DRIFT_EWMA_ALPHA = 0.35
DRIFT_FLAG_THRESHOLD = 0.5
HOTSWAP_CONSECUTIVE_THRESHOLD = 3

_FALLBACK_MODEL_CHAIN = ["primary-extractor-v1", "fallback-extractor-v2", "fallback-extractor-v3"]


class Sentinel:
    def __init__(self):
        self.events: list[SentinelEvent] = []

    def _next_fallback_signature(self, current: str) -> str:
        try:
            idx = _FALLBACK_MODEL_CHAIN.index(current)
        except ValueError:
            return _FALLBACK_MODEL_CHAIN[0]
        return _FALLBACK_MODEL_CHAIN[min(idx + 1, len(_FALLBACK_MODEL_CHAIN) - 1)]

    def evaluate(self, agent: AgentNode, anomaly: bool, reason: str) -> tuple[AgentNode, SentinelEvent | None]:
        """Returns the (possibly-replaced) agent plus an event if one fired."""
        sample = 1.0 if anomaly else 0.0
        agent.drift_score = round(DRIFT_EWMA_ALPHA * sample + (1 - DRIFT_EWMA_ALPHA) * agent.drift_score, 4)
        agent.consecutive_flags = agent.consecutive_flags + 1 if anomaly else 0

        if not anomaly:
            return agent, None

        if agent.consecutive_flags >= HOTSWAP_CONSECUTIVE_THRESHOLD:
            agent.status = AgentStatus.QUARANTINED
            replacement = AgentNode(
                agent_id=new_id("agent"),
                layer=agent.layer,
                role=agent.role,
                channel_type=agent.channel_type,
                sub_domain=agent.sub_domain,
                status=AgentStatus.ACTIVE,
                model_signature=self._next_fallback_signature(agent.model_signature),
                drift_score=0.0,
                consecutive_flags=0,
                leads_handled=0,
            )
            event = SentinelEvent(
                event_id=new_id("evt"),
                agent_id=agent.agent_id,
                event_type="HOTSWAP_TRIGGERED",
                drift_score=agent.drift_score,
                reason=f"{reason} (3 consecutive anomalies; quarantined and pinned to admin panel)",
                replacement_agent_id=replacement.agent_id,
            )
            self.events.append(event)
            return replacement, event

        if agent.drift_score >= DRIFT_FLAG_THRESHOLD:
            event = SentinelEvent(
                event_id=new_id("evt"),
                agent_id=agent.agent_id,
                event_type="DRIFT_FLAG",
                drift_score=agent.drift_score,
                reason=reason,
            )
            self.events.append(event)
            return agent, event

        return agent, None

File: iechm-leadgen-engine-c/src/sentinel.c
#include "sentinel.h"
#include <math.h>
#include <string.h>
#include <stdio.h>

static const double DRIFT_EWMA_ALPHA = 0.35;
static const double DRIFT_FLAG_THRESHOLD = 0.5;
static const int HOTSWAP_CONSECUTIVE_THRESHOLD = 3;

static const char *FALLBACK_CHAIN[] = {"primary-extractor-v1", "fallback-extractor-v2", "fallback-extractor-v3"};
#define FALLBACK_CHAIN_N 3

static double round4(double v) { return round(v * 10000.0) / 10000.0; }

static void next_fallback_signature(const char *current, char *out, size_t cap) {
    for (int i = 0; i < FALLBACK_CHAIN_N; i++) {
        if (strcmp(current, FALLBACK_CHAIN[i]) == 0) {
            int idx = i + 1 < FALLBACK_CHAIN_N ? i + 1 : FALLBACK_CHAIN_N - 1;
            xcpy(out, cap, FALLBACK_CHAIN[idx]);
            return;
        }
    }
    xcpy(out, cap, FALLBACK_CHAIN[0]);
}

void sentinel_init(Sentinel *s) { s->event_count = 0; }

SentinelEvalResult sentinel_evaluate(Sentinel *s, AgentNode *agent, bool anomaly, const char *reason) {
    SentinelEvalResult r;
    memset(&r, 0, sizeof r);

    double sample = anomaly ? 1.0 : 0.0;
    agent->drift_score = round4(DRIFT_EWMA_ALPHA * sample + (1.0 - DRIFT_EWMA_ALPHA) * agent->drift_score);
    agent->consecutive_flags = anomaly ? agent->consecutive_flags + 1 : 0;

    if (!anomaly) return r;

    if (agent->consecutive_flags >= HOTSWAP_CONSECUTIVE_THRESHOLD) {
        agent->status = AS_QUARANTINED;

        AgentNode replacement;
        memset(&replacement, 0, sizeof replacement);
        new_id(replacement.agent_id, sizeof replacement.agent_id, "agent");
        replacement.layer = agent->layer;
        xcpy(replacement.role, sizeof replacement.role, agent->role);
        replacement.has_channel_type = agent->has_channel_type;
        replacement.channel_type = agent->channel_type;
        replacement.has_sub_domain = agent->has_sub_domain;
        xcpy(replacement.sub_domain, sizeof replacement.sub_domain, agent->sub_domain);
        replacement.status = AS_ACTIVE;
        next_fallback_signature(agent->model_signature, replacement.model_signature, sizeof replacement.model_signature);
        replacement.drift_score = 0.0;
        replacement.consecutive_flags = 0;
        replacement.leads_handled = 0;
        now_iso(replacement.spawned_at, sizeof replacement.spawned_at);

        SentinelEvent ev;
        memset(&ev, 0, sizeof ev);
        new_id(ev.event_id, sizeof ev.event_id, "evt");
        xcpy(ev.agent_id, sizeof ev.agent_id, agent->agent_id);
        xcpy(ev.event_type, sizeof ev.event_type, "HOTSWAP_TRIGGERED");
        ev.drift_score = agent->drift_score;
        snprintf(ev.reason, sizeof ev.reason, "%s (3 consecutive anomalies; quarantined and pinned to admin panel)", reason);
        ev.has_replacement = true;
        xcpy(ev.replacement_agent_id, sizeof ev.replacement_agent_id, replacement.agent_id);
        now_iso(ev.created_at, sizeof ev.created_at);

        s->event_count++;
        r.replaced = true;
        r.replacement = replacement;
        r.has_event = true;
        r.event = ev;
        return r;
    }

    if (agent->drift_score >= DRIFT_FLAG_THRESHOLD) {
        SentinelEvent ev;
        memset(&ev, 0, sizeof ev);
        new_id(ev.event_id, sizeof ev.event_id, "evt");
        xcpy(ev.agent_id, sizeof ev.agent_id, agent->agent_id);
        xcpy(ev.event_type, sizeof ev.event_type, "DRIFT_FLAG");
        ev.drift_score = agent->drift_score;
        xcpy(ev.reason, sizeof ev.reason, reason);
        ev.has_replacement = false;
        now_iso(ev.created_at, sizeof ev.created_at);

        s->event_count++;
        r.has_event = true;
        r.event = ev;
        return r;
    }

    return r;
}
